File: sdk-go/audioutils.go
package llmsdk

import (
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"strings"
)

// base64ToInt16Samples converts base64 string to int16 samples
func base64ToInt16Samples(b64 string) ([]int16, error) {
	bytes, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return nil, fmt.Errorf("failed to decode base64: %w", err)
	}

	if len(bytes)%2 != 0 {
		return nil, fmt.Errorf("base64 data length is not a multiple of 2")
	}

	samples := make([]int16, len(bytes)/2)
	for i := range samples {
		samples[i] = int16(binary.LittleEndian.Uint16(bytes[i*2:]))
	}

	return samples, nil
}

// int16SamplesToBase64 converts int16 samples to base64 string
func int16SamplesToBase64(samples []int16) string {
	bytes := make([]byte, len(samples)*2)
	for i, sample := range samples {
		binary.LittleEndian.PutUint16(bytes[i*2:], uint16(sample))
	}
	return base64.StdEncoding.EncodeToString(bytes)
}

// concatenateB64AudioChunks concatenates audio data chunks into a single base64 string
func concatenateB64AudioChunks(chunks []string) (string, error) {
	if len(chunks) == 0 {
		return "", nil
	}

	var allSamples []int16
	for _, chunk := range chunks {
		samples, err := base64ToInt16Samples(chunk)
		if err != nil {
			return "", NewInvariantError("", fmt.Sprintf("Failed to decode audio chunk: %s", err.Error()))
		}
		allSamples = append(allSamples, samples...)
	}

	return int16SamplesToBase64(allSamples), nil
}

var audioFormatToMimeTypeMap = map[AudioFormat]string{
	AudioFormatWav:      "audio/wav",
	AudioFormatLinear16: "audio/L16",
	AudioFormatFLAC:     "audio/flac",
	AudioFormatMulaw:    "audio/basic",
	AudioFormatAlaw:     "audio/basic",
	AudioFormatMP3:      "audio/mpeg",
	AudioFormatOpus:     `audio/ogg; codecs="opus"`,
	AudioFormatAAC:      "audio/aac",
}

func MapAudioFormatToMimeType(format AudioFormat) string {
	if mimeType, ok := audioFormatToMimeTypeMap[format]; ok {
		return mimeType
	}
	return "application/octet-stream"
}

func MapMimeTypeToAudioFormat(mimeType string) (AudioFormat, error) {
	// strip out the parts after ;
	if idx := strings.Index(mimeType, ";"); idx != -1 {
		mimeType = strings.TrimSpace(mimeType[:idx])
	}
	for format, mt := range audioFormatToMimeTypeMap {
		if mimeType == mt {
			return format, nil
		}
	}
	return "", fmt.Errorf("unsupported audio format for mime type: %s", mimeType)
}

File: sdk-rust/src/utils/audio_utils.rs
use crate::{LanguageModelError, LanguageModelResult};
use base64::Engine as _;

pub fn base64_to_i16sample(b64: &str) -> Result<Vec<i16>, String> {
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(b64)
        .map_err(|e| format!("Failed to decode base64: {e}"))?;

    // Convert bytes to i16 samples (little-endian)
    if bytes.len() % 2 != 0 {
        return Err("Base64 data length is not a multiple of 2".to_string());
    }

    let mut samples = Vec::with_capacity(bytes.len() / 2);
    for chunk in bytes.chunks(2) {
        let sample = i16::from_le_bytes([chunk[0], chunk[1]]);
        samples.push(sample);
    }

    Ok(samples)
}

pub fn i16sample_to_base64(samples: &[i16]) -> String {
    // Convert i16 samples to bytes (little-endian)
    let mut result_bytes = Vec::with_capacity(samples.len() * 2);
    for &sample in samples {
        result_bytes.extend_from_slice(&sample.to_le_bytes());
    }

    // Encode to base64
    base64::engine::general_purpose::STANDARD.encode(result_bytes)
}

/// Concatenates audio data chunks into a single base64 string
pub fn concatenate_b64_audio_chunks(chunks: &[String]) -> LanguageModelResult<String> {
    if chunks.is_empty() {
        return Ok(String::new());
    }

    // Decode all chunks and collect samples
    let mut all_samples: Vec<i16> = Vec::new();

    for chunk in chunks {
        let samples = base64_to_i16sample(chunk).map_err(|e| {
            LanguageModelError::Invariant("", format!("Failed to decode audio chunk: {e}"))
        })?;
        all_samples.extend(samples);
    }

    let b64 = i16sample_to_base64(&all_samples);

    Ok(b64)
}
